workflow-runtime/src/main/java/com/squareup/workflow1/WorkflowInterceptor.kt

```diff
@@ -120,6 +120,103 @@ public interface WorkflowInterceptor {
     /** The parent [WorkflowSession] of this workflow, or null if this is the root workflow. */
     public val parent: WorkflowSession?
   }
+
+  /**
+   * A helper function intended for use by [onRender] methods that want to hook into action and
+   * side effect events.
+   *
+   * To use, wrap the [BaseRenderContext] with one of these, and override any of [onActionSent],
+   * [onSideEffectStarted].
+   *
+   * E.g.:
+   * ```
+   * override fun onRender(
+   *   renderProps: P,
+   *   renderState: S,
+   *   context: BaseRenderContext<P, S, O>,
+   *   proceed: (P, S, BaseRenderContext<P, S, O>) -> R,
+   *   session: WorkflowSession
+   * ): R = proceed(renderProps, renderState,
+   *   object : RenderContextInterceptor(context) {
+   *     override fun onActionSent(
+   *       action: WorkflowAction<P, S, O>,
+   *       proceed: (WorkflowAction<P, S, O>) -> Unit
+   *     ) {
+   *       log("Action sent: $action")
+   *       proceed(action)
+   *     }
+   *
+   *     override fun onSideEffectStarted(
+   *       key: String,
+   *       sideEffect: suspend CoroutineScope.() -> Unit,
+   *       proceed: (suspend CoroutineScope.() -> Unit) -> Unit
+   *     ) {
+   *       proceed {
+   *         log("Side effect started: $key")
+   *         try {
+   *           sideEffect()
+   *         } finally {
+   *           log("Side effect ended: $key")
+   *         }
+   *       }
+   *     }
+   *   }
+   * )
+   * ```
+   */
+  public abstract class RenderContextInterceptor<P, S, O>(
```

Comment:
Some problems with this approach:

1. Inheritance. Should this be an interface instead, or even just a function that takes some lambdas? Would eliminate the "don't call super" problem.
2. Requires allocating twice the number of context objects on each render pass than would otherwise be required. If we made these hooks more first-class (built them into `WorkflowInterceptor` directly, called them from `RealRenderContext`) then we could avoid that. But we're also allocating all these lambdas anyway, so what's one more object?

Reply:
Don't care even a tiny little bit about the lambdas.

I agree that an abstract class is awkward. Hard to choose between an interface or a factory function. Maybe an extension function on `BaseRenderContext`? Or even a couple / three, for tidier syntax if you only want to use one method or the other.

Reply:
I don't like the extension function syntax since it will pollute the autocomplete namespace for all workflows' `render` methods that have even transient dependencies on the runtime. If we had multiple receivers, I could do an extension on the union of (`WorkflowInterceptor & BaseRenderContext`).

Reply:
I don't like putting these functions on the main `Workflowinterceptor` interface because then they don't have access to any of the state/props from the render pass that the context was created for, which could be useful to associate.